README.md
# PHP Autoloader
Lightweight PSR-4 PHP Autoloader Class.

[![Latest Stable Version](https://poser.pugx.org/varunsridharan/php-autoloader/version)](https://packagist.org/packages/varunsridharan/php-autoloader)
[![Total Downloads](https://poser.pugx.org/varunsridharan/php-autoloader/downloads)](https://packagist.org/packages/varunsridharan/php-autoloader)
[![Latest Unstable Version](https://poser.pugx.org/varunsridharan/php-autoloader/v/unstable)](//packagist.org/packages/varunsridharan/php-autoloader)
[![License](https://poser.pugx.org/varunsridharan/php-autoloader/license)](https://packagist.org/packages/varunsridharan/php-autoloader)
[![composer.lock available](https://poser.pugx.org/varunsridharan/php-autoloader/composerlock)](https://packagist.org/packages/varunsridharan/php-autoloader) 


## Installation
The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

To install **PHP_Autoloader library**, simply:

    $ composer require varunsridharan/php-autoloader

The previous command will only install the necessary files, if you prefer to **download the entire source code** you can use:

    $ composer require varunsridharan/php-autoloader --prefer-source

You can also **clone the complete repository** with Git:

    $ git clone https://github.com/varunsridharan/php-autoloader.git

Or **install it manually**:

[Download autoloader.php](https://raw.githubusercontent.com/varunsridharan/php-autoloader/master/src/autoloader.php):

    $ wget https://raw.githubusercontent.com/varunsridharan/php-autoloader/master/src/autoloader.php


## Usage
### Arguments
* **`$namespace` :** PHP Namspace To Handle Autoload For  
* **`$classmap` :** https://github.com/varunsridharan/php-classmap-generator Generated Classmap File.
* **`$path` :** Full Path to lookup for php files
* **`$options` :** An Array of useful Arguments.
* **`$prepend` :** Option to prepend / append current autoloader with exists autoloaders.

#### `$options` Arguments
```php
  array(
      /**
       * An Array of Class Namespace to Exclude While Checking for current namespace.
       * Eg
       * Main Namespace \Testing\Core
       * Exclude \Testing\Core\Abstracts
       */
      'exclude' => false,
      
      /**
       * Custom Option To Quickly Remap A Class File. 
       * If any class added then it will not search.
       * instead it gets the location from here and loads it.
       * An Array of class and its file location
       */
      'mapping' => array(),
      
      /**
       * Set To True / False.
       */
      'debug'   => false,
  );
```

### Example Folder Stucture
```php
  | - src\
  | -- class1\
  | --- class1.php
  | -- class2\
  | --- class2.php
  | --- class3\
  | ---- class3.php
  | - loader.php
  | - index.php
```
### `loader.php` SourceCode
```php
  $autoloader = new \Varunsridharan\PHP\Autoloader('varun',__DIR__.'/src/',array(
      'mapping' => array(
          'varun\class2\class3\class3' => 'src/class2/class3/class3.php',
      )
  ))
```

### `index.php` SoruceCode
```php
  require __DIR__.'loader.php';
  
  $class1 = new \varun\class1\class1(); // This file is autoloaded based on the namespace 
  $class2 = new \varun\class1\class2(); // This file is autoloaded based on the namespace 
  $class3 = new \varun\class1\class3(); // This file is loaded using the data from remap array
```


---

<!-- START common-footer.mustache  -->
## 📝 Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

[Checkout CHANGELOG.md](https://github.com/varunsridharan/php-autoloader/blob/main/CHANGELOG.md)


## 🤝 Contributing
If you would like to help, please take a look at the list of [issues](https://github.com/varunsridharan/php-autoloader/issues/).


## 📜  License & Conduct
- [**GNU General Public License v3.0**](https://github.com/varunsridharan/php-autoloader/blob/main/LICENSE) © [Varun Sridharan](website)
- [Code of Conduct](https://github.com/varunsridharan/.github/blob/main/CODE_OF_CONDUCT.md)


## 📣 Feedback
- ⭐ This repository if this project helped you! :wink:
- Create An [🔧 Issue](https://github.com/varunsridharan/php-autoloader/issues/) if you need help / found a bug


## 💰 Sponsor
[I][twitter] fell in love with open-source in 2013 and there has been no looking back since! You can read more about me [here][website].
If you, or your company, use any of my projects or like what I’m doing, kindly consider backing me. I'm in this for the long run.

- ☕ How about we get to know each other over coffee? Buy me a cup for just [**$9.99**][buymeacoffee]
- ☕️☕️ How about buying me just 2 cups of coffee each month? You can do that for as little as [**$9.99**][buymeacoffee]
- 🔰         We love bettering open-source projects. Support 1-hour of open-source maintenance for [**$24.99 one-time?**][paypal]
- 🚀         Love open-source tools? Me too! How about supporting one hour of open-source development for just [**$49.99 one-time ?**][paypal]

<!-- Personl Links -->
[paypal]: https://sva.onl/paypal
[buymeacoffee]: https://sva.onl/buymeacoffee
[twitter]: https://sva.onl/twitter/
[website]: https://sva.onl/website/


## Connect & Say 👋
- **Follow** me on [👨‍💻 Github][github] and stay updated on free and open-source software
- **Follow** me on [🐦 Twitter][twitter] to get updates on my latest open source projects
- **Message** me on [📠 Telegram][telegram]
- **Follow** my pet on [Instagram][sofythelabrador] for some _dog-tastic_ updates!

<!-- Personl Links -->
[sofythelabrador]: https://www.instagram.com/sofythelabrador/
[github]: https://sva.onl/github/
[twitter]: https://sva.onl/twitter/
[telegram]: https://sva.onl/telegram/


---

<p align="center">
<i>Built With ♥ By <a href="https://sva.onl/twitter"  target="_blank" rel="noopener noreferrer">Varun Sridharan</a> <a href="https://en.wikipedia.org/wiki/India">
   <img src="https://cdn.svarun.dev/flag-india.jpg" width="20px"/></a> </i> <br/><br/>
   <img src="https://cdn.svarun.dev/codeispoetry.png"/>
</p>

---


<!-- END common-footer.mustache  -->


CHANGELOG.md
# 📝  Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 2.6 - 04/10/2019
### Fixed
* Improved Classmap Autoloader.

## 2.5 - 02/10/2019
### Added
* Option to use class map file generated by https://github.com/varunsridharan/php-classmap-generator

## 2.4 - 14/05/2019
### Fixed
* Minor Bug

## 2.3 - 13/05/2019
### Fixed
* Minor Bug

## 2.2 - 09/05/2019
### Fixed
* Minor Bug

## 2.1 - 09/05/2019
### Fixed
* Minor Bug

## 2.0 - 07/05/2019
### Built From Scratch

## 1.4 - 07/05/2019
### Added
* `class_exists` checks

## 1.3.3 - 24/04/2019
### Fixed
* Array Notice

## 1.3.2 - 31/03/2019
### Fixed
* Minor Bug

## 1.3.1 - 02/02/2019
### Fixed
*  Issue with loader in `search_folders` fixed.

## 1.3 - 14/01/2019
### Fixed
* Issue With Auto loader Fixed.

## 1.2 - 11/01/2019
### Added
* Option To Exclude Certain Namspace
* Option To Lookup for class files with more vast lookup options

## 1.1 - 21/12/2018
### Changed
* Updated Composer.json With Proper Autoloader Values.

## 1.0 - 21/12/2018
### First Release


<!--
## Unreleased

## 1.0 - 01/02/2020
### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

-->


src/autoloader.php
<?php
/**
 * Lightweight PSR-4 PHP Autoloader Class.
 *
 * @author Varun Sridharan <varunsridharan23@gmail.com>
 * @version 1.0
 * @since 1.0
 * @link
 * @copyright 2018 Varun Sridharan
 * @license GPLV3 Or Greater (https://www.gnu.org/licenses/gpl-3.0.txt)
 */

namespace Varunsridharan\PHP;

/**
 * Class Autoloader
 *
 * @package Varunsridharan\PHP
 * @author Varun Sridharan <varunsridharan23@gmail.com>
 * @since 1.0
 */
class Autoloader {
	/**
	 * @var null
	 * @access
	 */
	protected $namespace = null;

	/**
	 * @var null
	 * @access
	 */
	protected $base_path = null;

	/**
	 * @var null
	 * @access
	 */
	protected $mapping = null;

	/**
	 * Search And Stores All Subfolders in the base dir if required.
	 *
	 * @var array
	 * @access protected
	 */
	protected $subfolders = array();

	/**
	 * Stores Debug Information.
	 *
	 * @var array
	 * @access
	 */
	protected $debug = array();

	/**
	 * Stores Class Operation.
	 *
	 * @var array
	 * @access
	 */
	protected $options = array();

	/**
	 * Stores Current Lookup Class.
	 *
	 * @var bool
	 * @access
	 */
	protected $lookup_class = false;

	/**
	 * Autoloader constructor.
	 *
	 * @param string      $namespace
	 * @param string|bool $path
	 * @param array       $options
	 * @param             $prepend_autoload
	 *
	 * @throws \Exception
	 */
	public function __construct( $namespace = '', $path = false, $options = array(), $prepend_autoload = false ) {
		if ( ! is_array( $namespace ) ) {
			$this->namespace = $namespace;
			$this->base_path = $this->trailingslashit( $path );

			$default_options = array(
				'exclude' => false,
				'mapping' => array(),
				'debug'   => false,
			);
			$this->options   = array_merge( $default_options, $options );
			if ( false !== $this->options['exclude'] && ! is_array( $this->options['exclude'] ) ) {
				$this->options['exclude'] = array( $this->options['exclude'] );
			}

			$this->mapping = $this->options['mapping'];
			$this->register( $prepend_autoload );
			unset( $this->options['mapping'] );
		}
	}

	/**
	 * @param      $key
	 * @param bool $default
	 *
	 * @return bool|mixed
	 */
	protected function option( $key, $default = false ) {
		return ( isset( $this->options[ $key ] ) ) ? $this->options[ $key ] : $default;
	}

	/**
	 * Registers Autoloader With PHP.
	 *
	 * @param bool $prepend
	 *
	 * @return $this
	 * @throws \Exception
	 */
	public function register( $prepend = false ) {
		spl_autoload_register( array( &$this, 'load_class' ), true, $prepend );
		return $this;
	}

	/**
	 * De-Registers Autoloader In PHP.
	 *
	 * @return $this
	 */
	public function unregister() {
		spl_autoload_unregister( array( &$this, 'load_class' ) );
		return $this;
	}

	/**
	 * Removes trailing forward slashes and backslashes if they exist.
	 *
	 * The primary use of this is for paths and thus should be used for paths. It is
	 * not restricted to paths and offers no specific path support.
	 *
	 * @param string $string What to remove the trailing slashes from.
	 *
	 * @return string String without the trailing slashes.
	 */
	protected function untrailingslashit( $string ) {
		return rtrim( $string, '/\\' );
	}

	/**
	 * Appends a trailing slash.
	 *
	 * Will remove trailing forward and backslashes if it exists already before adding
	 * a trailing forward slash. This prevents double slashing a string or path.
	 *
	 * The primary use of this is for paths and thus should be used for paths. It is
	 * not restricted to paths and offers no specific path support.
	 *
	 * @param string $string What to add the trailing slash to.
	 *
	 * @return string String with trailing slash added.
	 */
	protected function trailingslashit( $string ) {
		return $this->untrailingslashit( $string ) . '/';
	}

	/**
	 * Stores Debug Message.
	 *
	 * @param $msg
	 *
	 * @return $this
	 */
	protected function debug( $msg ) {
		if ( true === $this->option( 'debug' ) ) {
			$this->debug[] = $msg;
		} elseif ( $this->lookup_class === $this->option( 'debug' ) ) {
			$this->debug[] = $msg;
		}
		return $this;
	}

	/**
	 * Returns Debug Informations.
	 *
	 * @return array
	 */
	public function show_debug() {
		return $this->debug;
	}

	/**
	 * Validates If Its A Proper Lookup Namespace.
	 *
	 * @return bool
	 */
	public function is_valid_lookup() {
		if ( strpos( strtolower( $this->lookup_class ), strtolower( $this->namespace ) ) !== false ) {
			if ( false !== $this->option( 'exclude' ) ) {
				foreach ( $this->option( 'exclude' ) as $namespace ) {
					if ( strpos( strtolower( $this->lookup_class ), strtolower( $namespace ) ) !== false ) {
						return false;
					}
				}
			}
			return true;
		}
		return false;
	}

	/**
	 * Works For PHP SPL Autoloader and loads class based on the requests.
	 *
	 * @param $org_class
	 */
	public function load_class( $org_class ) {
		$this->lookup_class = $org_class;
		if ( $this->is_valid_lookup() ) {
			$filenames = array();
			$folders   = array();
			$is_loaded = false;

			$this->debug( 'Class Requested : ' . $org_class );
			$this->debug( 'Checking in Mapping Array' );

			if ( isset( $this->mapping[ $org_class ] ) && ( file_exists( $this->mapping[ $org_class ] ) || file_exists( $this->base_path . $this->mapping[ $org_class ] ) ) ) {
				$is_loaded = $this->load_file( $this->mapping[ $org_class ], $org_class );

				if ( false === $is_loaded ) {
					$is_loaded = $this->load_file( $this->base_path . $this->mapping[ $org_class ], $org_class );
				}
			}

			if ( false === $is_loaded ) {
				$filenames = $this->get_file_name( $org_class );
				$folders   = $this->get_folder_name( $org_class );
				$this->debug( 'Not Found In Remap Array' );
			}

			if ( false === $is_loaded ) {
				$this->debug( '' );
				$this->debug( 'Checking Using Requested Class Namespace' );
				if ( is_array( $filenames ) && is_array( $folders ) ) {
					foreach ( $folders as $folder ) {
						foreach ( $filenames as $file ) {
							$is_loaded = $this->load_file( $this->trailingslashit( $this->base_path . $folder ) . $file, $org_class );
							if ( $is_loaded ) {
								$this->debug( 'Class Found' );
								break;
							} else {
								$this->debug( 'File Not Found In : ' . $this->trailingslashit( $this->base_path . $folder ) . $file );
							}
						}
						if ( $is_loaded ) {
							break;
						}
					}
				}
			}

			if ( false === $is_loaded ) {
				$this->debug( '' );
				$this->debug( 'Checking In All subfolders Inside Base Path' );
				$folders = $this->get_folders( $this->base_path );
				foreach ( $folders as $folder ) {
					foreach ( $filenames as $file ) {
						$is_loaded = $this->load_file( $this->trailingslashit( $folder ) . $file, $org_class );
						if ( $is_loaded ) {
							$this->debug( 'Class Found' );
							break;
						} else {
							$this->debug( 'File Not Found In : ' . $this->trailingslashit( $folder ) . $file );
						}
					}
				}
			}

			$this->debug( '' );
		}

		$this->lookup_class = false;
	}

	/**
	 * @param $path
	 * @param $class
	 *
	 * @return bool
	 */
	protected function load_file( $path, $class ) {
		if ( file_exists( $path ) ) {
			require_once $path;
			return ( class_exists( $class, false ) );
		}
		return false;
	}

	/**
	 * Saves & Returns Core Folders.
	 *
	 * @param $path
	 *
	 * @return mixed
	 */
	protected function get_folders( $path ) {
		if ( isset( $this->subfolders[ $path ] ) ) {
			return $this->subfolders[ $path ];
		}

		$this->subfolders[ $path ] = $this->search_folders( $path );
		return $this->subfolders[ $path ];
	}

	/**
	 * Search folders & subfolders inside a given folder and returns it.
	 *
	 * @param $path
	 *
	 * @return array|string
	 */
	protected function search_folders( $path ) {
		$return       = array();
		$base_folders = array_filter( glob( $this->trailingslashit( $path ) . '*' ), 'is_dir' );

		if ( is_array( $base_folders ) && ! empty( $base_folders ) ) {
			foreach ( $base_folders as $folder ) {
				$_sub = $this->search_folders( $this->untrailingslashit( $folder ) );

				if ( is_array( $_sub ) && ! empty( $_sub ) ) {
					$return = array_merge( $return, $_sub );
				} elseif ( is_string( $_sub ) ) {
					$return = array_merge( $return, array( $_sub ) );
					$return = array_merge( $return, array( $this->untrailingslashit( $path ) ) );
				}
			}
		} else {
			$return = $this->trailingslashit( $path );
		}

		return $return;
	}

	/**
	 * Creates File Names From Class Name.
	 *
	 * @param $class
	 *
	 * @return array|bool
	 */
	protected function get_file_name( $class ) {
		if ( $class ) {
			$file   = explode( '\\', $class );
			$file   = strtolower( end( $file ) );
			$file__ = str_replace( '_', '-', $file );
			$files  = array(
				$file . '.php', // File Name With _.
				$file__ . '.php', // File Name With - which is replaced by _.
				'class-' . $file__ . '.php', // File name With Class Prefixed.
				'class_' . $file__ . '.php', // File Name With - which is replaced by _.
			);
			return $files;
		}
		return false;
	}

	/**
	 * @param $class
	 *
	 * @return array|bool
	 */
	protected function get_folder_name( $class ) {
		if ( $class ) {
			$class = str_replace( $this->namespace, '', $class );
			$file  = explode( '\\', $class );
			$file  = end( $file );
			$class = str_replace( $file, '', $class );
			$class = strtolower( str_replace( '\\', '/', $class ) );

			return array(
				$class,
				str_replace( '_', '-', $class ),
				strtolower( $file ),
				str_replace( '-', '_', $class ),
			);
		}
		return false;
	}

	/**
	 * Adds Custom mapping To mapping Array.
	 *
	 * @param $class
	 * @param $file
	 *
	 * @return $this
	 */
	public function add( $class, $file ) {
		$this->mapping[ $class ] = $file;
		return $this;
	}

	/**
	 * Removes Custom mapping To a Class in mapping Array.
	 *
	 * @param $classs
	 *
	 * @return $this
	 */
	public function remove( $classs ) {
		if ( isset( $this->mapping[ $classs ] ) ) {
			unset( $this->mapping[ $classs ] );
		}
		return $this;
	}
}


src/Autoloader.php
<?php
/**
 * Lightweight PSR-4 PHP Autoloader Class.
 *
 * @author Varun Sridharan <varunsridharan23@gmail.com>
 * @version 1.0
 * @since 1.0
 * @copyright 2018 Varun Sridharan
 * @license GPLV3 Or Greater (https://www.gnu.org/licenses/gpl-3.0.txt)
 */

namespace Varunsridharan\PHP;

use Exception;

if ( ! class_exists( '\Varunsridharan\PHP\Autoloader' ) ) {
	/**
	 * Class Autoloader
	 *
	 * @package Varunsridharan\PHP
	 * @author Varun Sridharan <varunsridharan23@gmail.com>
	 * @since 1.0
	 */
	class Autoloader {
		/**
		 * Stores Subfolder Informations.
		 *
		 * @var array
		 */
		protected static $subfolders = array();

		/**
		 * @var array
		 */
		protected $options = array();

		/**
		 * Stores Namespace.
		 *
		 * @var null
		 */
		protected $namespace = null;

		/**
		 * Stores Base Path.
		 *
		 * @var bool
		 */
		protected $base_path = false;

		/**
		 * Stores Mapping.
		 * Stores an array of class name and file as key value in a array
		 *
		 * @example array(
		 *    '\somenamespace\someclass' => 'your-path/file.php'
		 *    '\somenamespace\someclass2' => 'your-path/file2.php'
		 * )
		 *
		 * @var bool
		 */
		protected $mappings = false;

		/**
		 * Stores Classmap Generated Files.
		 *
		 * @var bool
		 */
		protected $classmap = false;

		/**
		 * Stores All Classmaps.
		 *
		 * @var array
		 */
		protected static $classmaps = array();

		/**
		 * Autoloader constructor.
		 *
		 * @param $namespace
		 * @param $path
		 * @param $options
		 */
		public function __construct( $namespace, $path, $options = array() ) {
			$this->set_options( $options );

			if ( $this->register() ) {
				$this->namespace = $namespace;
				$this->base_path = $path;
			}
		}

		/**
		 * Sets Option.
		 *
		 * @param $options
		 *
		 * @return $this
		 */
		protected function set_options( $options ) {
			$this->options            = array_merge( array(
				'exclude'        => false,
				'prepend'        => false,
				'mapping'        => false,
				'search_folders' => false,
				'classmap'       => false,
			), $options );
			$this->options['exclude'] = ( ! is_array( $this->options['exclude'] ) ) ? array_filter( array( $this->options['exclude'] ) ) : $this->options['exclude'];
			$this->options['mapping'] = ( ! is_array( $this->options['mapping'] ) ) ? array() : $this->options['mapping'];
			$this->mappings           = $this->options['mapping'];

			if ( ! empty( $this->options['classmap'] ) && is_string( $this->options['classmap'] ) && file_exists( $this->options['classmap'] ) ) {
				$this->classmap = $this->options['classmap'];
				if ( ! isset( self::$classmaps[ $this->options['classmap'] ] ) ) {
					$data = include $this->options['classmap'];
					foreach ( $data as $key => $value ) {
						unset( $data[ $key ] );
						$data[ strtolower( $key ) ] = $value;
					}
					self::$classmaps[ $this->options['classmap'] ] = $data;
				}
			}

			unset( $this->options['mapping'] );
			unset( $this->options['classmaps'] );
			return $this;
		}

		/**
		 * Returns An Option.
		 *
		 * @param      $key
		 * @param bool $default
		 *
		 * @return bool|mixed
		 */
		protected function option( $key, $default = false ) {
			return ( isset( $this->options[ $key ] ) ) ? $this->options[ $key ] : $default;
		}

		/**
		 * Registers Autoloader With PHP.
		 *
		 * @return bool
		 */
		protected function register() {
			try {
				spl_autoload_register( array( &$this, 'autoload' ), true, $this->option( 'prepend' ) );
			} catch ( Exception $exception ) {
				return false;
			}
			return true;
		}

		/**
		 * De-Registers Autoloader In PHP.
		 *
		 * @return bool|boolean
		 */
		public function unregister() {
			return spl_autoload_unregister( array( &$this, 'autoload' ) );
		}

		/**
		 * Removes trailing forward slashes and backslashes if they exist.
		 *
		 * @param string $string What to remove the trailing slashes from.
		 *
		 * @return string String without the trailing slashes.
		 */
		protected function unslashit( $string ) {
			return rtrim( $string, '/\\' );
		}

		/**
		 * Appends a trailing slash.
		 *
		 * @param string $string What to add the trailing slash to.
		 *
		 * @return string String with trailing slash added.
		 */
		protected function slashit( $string ) {
			return $this->unslashit( $string ) . '/';
		}

		/**
		 * @param $class
		 */
		public function autoload( $class ) {
			if ( true === $this->is_valid_lookup( $class ) ) {
				$lclass    = strtolower( $class );
				$ckey      = $this->classmap;
				$filenames = null;
				$folders   = null;
				$is_loaded = false;

				/**
				 * @uses varunsridharan/php-classmap-generator script which provides location along with class name
				 */
				if ( isset( self::$classmaps[ $ckey ][ $lclass ] ) && file_exists( $this->base_path . self::$classmaps[ $ckey ][ $lclass ] ) ) {
					$this->load_file( $this->base_path . self::$classmaps[ $ckey ][ $lclass ], $class );
				} else {
					/**
					 * Checks and loads file if given class exists in mapping array.
					 *
					 * @example array(
					 *    '\somenamespace\someclass' => 'your-path/file.php'
					 *    '\somenamespace\someclass2' => 'your-path/file2.php'
					 * )
					 */
					if ( false === $is_loaded && isset( $this->mappings[ $class ] ) ) {
						$is_loaded = $this->load_file( $this->mappings[ $class ], $class );
						if ( false === $is_loaded ) {
							$is_loaded = $this->load_file( $this->base_path . $this->mappings[ $class ], $class );
						}
					}

					/**
					 * Checks and loads class based on the files & folder found.
					 */
					if ( false === $is_loaded ) {
						$filenames = $this->get_possible_filenames( $class );
						$folders   = $this->get_possible_foldernames( $class );

						if ( is_array( $filenames ) && is_array( $folders ) && ! empty( $filenames ) && ! empty( $folders ) ) {
							foreach ( $folders as $folder ) {
								foreach ( $filenames as $file ) {
									$is_loaded = $this->load_file( $this->slashit( $this->base_path . $folder ) . $file, $class );
									if ( $is_loaded ) {
										break;
									}
								}
								if ( $is_loaded ) {
									break;
								}
							}
						}
					}

					/**
					 * Checks and loads class based on all subfolder in the given path
					 */
					if ( false === $is_loaded && true === $this->option( 'search_folders' ) ) {
						$folders = $this->get_folders( $this->base_path );

						if ( is_array( $folders ) && is_array( $filenames ) && ! empty( $filenames ) ) {
							foreach ( $folders as $folder ) {
								foreach ( $filenames as $file ) {
									$is_loaded = $this->load_file( $this->slashit( $folder ) . $file, $class );
									if ( $is_loaded ) {
										break;
									}
								}
							}
						}
					}
				}
			}
		}

		/**
		 * Saves & Returns Core Folders.
		 *
		 * @param $path
		 *
		 * @return mixed
		 */
		protected function get_folders( $path ) {
			if ( isset( self::$subfolders[ $path ] ) ) {
				return self::$subfolders[ $path ];
			}
			self::$subfolders[ $path ] = array_unique( array_filter( $this->search_folders( $path ) ) );
			return self::$subfolders[ $path ];
		}

		/**
		 * Search folders & subfolders inside a given folder and returns it.
		 *
		 * @param null $path
		 *
		 * @return array|string
		 */
		protected function search_folders( $path = null ) {
			$return       = ( null === $path || $path === $this->base_path ) ? array( $this->base_path ) : array();
			$path         = ( null === $path ) ? $this->base_path : $path;
			$base_folders = array_filter( glob( $this->slashit( $path ) . '*', GLOB_ONLYDIR ) );

			if ( ! empty( $base_folders ) ) {
				foreach ( $base_folders as $folder ) {
					$return[] = $folder;
					$return   = array_merge( $return, $this->search_folders( $folder ) );
				}
			}
			return $return;
		}

		/**
		 * Returns Actual Class Name.
		 *
		 * @param $class
		 *
		 * @return string
		 * @example $class = \YourClassNameSpace\AnotherNamespace\YourClass then it returns as yourclass
		 */
		protected function get_actual_classname( $class ) {
			$file = explode( '\\', strtolower( $class ) );
			return strtolower( end( $file ) );
		}

		/**
		 * Returns Possible Filenames.
		 *
		 * @param $class
		 *
		 * @return array|bool
		 */
		protected function get_possible_filenames( $class ) {
			if ( ! empty( $class ) ) {
				$class_name = $this->get_actual_classname( $class );
				if ( empty( $class_name ) ) {
					return false;
				}
				// Replaces Strings that has underscores with - (your_string to your-string)
				$class_with_hypen = str_replace( '_', '-', $class_name );
				// Replaces Strings that has underscores with - (your_string to your.string)
				$class_with_dot = str_replace( '_', '.', $class_name );

				/**
				 * @example array (
				 *    'your_class.php',
				 *    'your-class.php',
				 *    'your.class.php',
				 *    'class_your_class.php',
				 *    'class-your-class.php',
				 *    'class.your.class.php',
				 *    'abstract_your_class.php',
				 *    'abstract-your-class.php',
				 *    'abstract.your.class.php',
				 *    'trait_your_class.php',
				 *    'trait-your-class.php',
				 *    'trait.your.class.php',
				 *    'interface_your_class.php',
				 *    'interface-your-class.php',
				 *    'interface.your.class.php',
				 * )
				 */
				$return = array();

				$return[] = 'class-' . $class_with_hypen . '.php';
				$return[] = $class_with_hypen . '.php';
				$return[] = 'abstract-' . $class_with_hypen . '.php';
				$return[] = 'trait-' . $class_with_hypen . '.php';
				$return[] = 'interface-' . $class_with_hypen . '.php';

				$return[] = 'class_' . $class_name . '.php';
				$return[] = $class_name . '.php';
				$return[] = 'abstract_' . $class_name . '.php';
				$return[] = 'trait_' . $class_name . '.php';
				$return[] = 'interface_' . $class_name . '.php';

				$return[] = 'class.' . $class_with_dot . '.php';
				$return[] = $class_with_dot . '.php';
				$return[] = 'abstract.' . $class_with_dot . '.php';
				$return[] = 'trait.' . $class_with_dot . '.php';
				$return[] = 'interface.' . $class_with_dot . '.php';

				return array_unique( array_filter( $return ) );
			}
			return false;
		}

		/**
		 * @param $class
		 *
		 * @return array|boolean
		 *
		 * @example array (size=3)
		 *    1 => string '/path1/path1_base/path2/' (length=24)
		 *    2 => string '/path1/path1-base/path2/' (length=24)
		 *    3 => string 'your_class' (length=10)
		 *    4 => string 'your-class' (length=10)
		 * )
		 */
		protected function get_possible_foldernames( $class ) {
			$class_name = $this->get_actual_classname( $class );
			if ( empty( $class_name ) ) {
				return false;
			}

			$_class = explode( '\\', $class );

			$_keys = array_keys( $_class );
			if ( ! empty( $_keys ) ) {
				$end = end( $_keys );
				if ( ! empty( $end ) ) {
					unset( $_class[ $end ] );
				}
			}
			$_class = strtolower( implode( '\\', $_class ) );

			$folder_class = str_replace( strtolower( $this->namespace ), '', $_class );
			$folder_class = str_replace( '\\', '/', $folder_class );
			$folder_class = trim( $folder_class, '/' );
			$folder_class = trim( $folder_class, '\\' );

			return array_unique( array_filter( array(
				$folder_class,
				str_replace( '_', '-', $folder_class ),
				$class_name,
				str_replace( '_', '-', $class_name ),
				'/',
			) ) );
		}

		/**
		 * @param $path
		 * @param $class
		 *
		 * @return bool
		 */
		protected function load_file( $path, $class ) {
			if ( is_readable( $path ) ) {
				include_once $path;
			}
			return ( class_exists( $class, false ) || trait_exists( $class, false ) || interface_exists( $class, false ) );
		}

		/**
		 * Checks if requested class is a valid lookup.
		 *
		 * @param string $class
		 *
		 * @return bool
		 */
		protected function is_valid_lookup( $class ) {
			$namespace = strtolower( $this->namespace );
			$lookup    = strtolower( $class );

			if ( false === strpos( $lookup, $namespace, false ) ) {
				return false;
			}

			if ( ! empty( $this->option( 'exclude' ) ) && is_array( $this->option( 'exclude' ) ) ) {
				foreach ( $this->option( 'exclude' ) as $namespace ) {
					$namespace = strtolower( $namespace );
					if ( false !== strpos( $lookup, strtolower( $namespace ) ) || false !== strpos( '\\' . $lookup, strtolower( $namespace ) ) ) {
						return false;
					}
				}
			}
			return true;
		}

		/**
		 * Adds Mapping.
		 *
		 * @param $class
		 * @param $file
		 *
		 * @return $this
		 */
		public function map( $class, $file ) {
			$this->mappings[ $class ] = $file;
			return $this;
		}

		/**
		 * Unmaps.
		 *
		 * @param $class
		 *
		 * @return $this
		 */
		public function unmap( $class ) {
			if ( isset( $this->mappings[ $class ] ) ) {
				unset( $this->mappings[ $class ] );
			}
			return $this;
		}
	}
}
